File: internal/daemon/uds/uds.go
// Package uds wraps net.Listen("unix", ...) with single-instance semantics
// suitable for the autosk daemon socket.
//
// Listen ensures the parent directory exists with mode 0700, refuses to
// take over a socket whose other end accepts connections (a live peer),
// unlinks stale leftovers, binds, and chmods the resulting socket to
// 0600. Cleanup deletes the socket on shutdown.
//
// Per docs/plans/20260518-Daemon-UDS-Plan.md §4.3.
package uds

import (
	"errors"
	"fmt"
	"net"
	"os"
	"path/filepath"
	"syscall"
	"time"
)

// ErrAlreadyRunning is returned when the socket path already has a live peer.
var ErrAlreadyRunning = errors.New("uds: daemon already running")

// dialTimeout is the time we give a stale-vs-live probe before deciding
// the existing socket has no listener.
const dialTimeout = 200 * time.Millisecond

// Listen returns a unix-domain listener bound at path with strict single-
// instance semantics:
//
//  1. filepath.Dir(path) is created with mode 0700 (best-effort).
//  2. If a file already exists at path:
//     - net.Dial succeeds  → ErrAlreadyRunning.
//     - dial fails with ECONNREFUSED/ENOENT → unlink + bind.
//  3. After net.Listen succeeds, os.Chmod(path, 0o600).
//
// Any error from os.Remove other than ENOENT short-circuits — we never
// blindly clobber a path we cannot identify as our own stale socket.
func Listen(path string) (net.Listener, error) {
	if path == "" {
		return nil, errors.New("uds: empty socket path")
	}
	dir := filepath.Dir(path)
	if dir != "" && dir != "." {
		// Best-effort: if the directory already exists with a different
		// mode we don't try to chmod it (mDirs created by the user are
		// out of our scope; we only own the socket file itself).
		if err := os.MkdirAll(dir, 0o700); err != nil {
			return nil, fmt.Errorf("uds: mkdir %s: %w", dir, err)
		}
	}

	if _, err := os.Lstat(path); err == nil {
		// Path exists. Decide: live peer or stale leftover?
		c, derr := net.DialTimeout("unix", path, dialTimeout)
		if derr == nil {
			_ = c.Close()
			return nil, fmt.Errorf("%w at %s", ErrAlreadyRunning, path)
		}
		// Not a live peer — remove and continue.
		if !isDeadSocketErr(derr) {
			// Some other dial error (permission denied etc.) — surface so
			// the user can fix the root cause.
			return nil, fmt.Errorf("uds: probe %s: %w", path, derr)
		}
		if rerr := os.Remove(path); rerr != nil && !errors.Is(rerr, os.ErrNotExist) {
			return nil, fmt.Errorf("uds: remove stale socket %s: %w", path, rerr)
		}
	} else if !errors.Is(err, os.ErrNotExist) {
		return nil, fmt.Errorf("uds: stat %s: %w", path, err)
	}

	ln, err := net.Listen("unix", path)
	if err != nil {
		return nil, fmt.Errorf("uds: listen %s: %w", path, err)
	}
	if cerr := os.Chmod(path, 0o600); cerr != nil {
		_ = ln.Close()
		_ = os.Remove(path)
		return nil, fmt.Errorf("uds: chmod %s: %w", path, cerr)
	}
	return ln, nil
}

// Cleanup removes the socket file at path. Errors other than ENOENT are
// surfaced; ENOENT is treated as a no-op.
func Cleanup(path string) error {
	if path == "" {
		return nil
	}
	if err := os.Remove(path); err != nil && !errors.Is(err, os.ErrNotExist) {
		return err
	}
	return nil
}

// isDeadSocketErr reports whether a Dial error is consistent with a
// stale leftover (no peer accepting connections, or socket file missing
// between Stat and Dial).
func isDeadSocketErr(err error) bool {
	if err == nil {
		return false
	}
	if errors.Is(err, syscall.ECONNREFUSED) ||
		errors.Is(err, syscall.ECONNRESET) ||
		errors.Is(err, syscall.ENOENT) {
		return true
	}
	// net.OpError sometimes wraps a *os.PathError or a generic string.
	var opErr *net.OpError
	if errors.As(err, &opErr) {
		if errors.Is(opErr.Err, syscall.ECONNREFUSED) ||
			errors.Is(opErr.Err, syscall.ECONNRESET) ||
			errors.Is(opErr.Err, syscall.ENOENT) {
			return true
		}
	}
	return false
}

File: crates/autoskd/src/uds.rs
//! Unix-domain-socket single-instance binding — the Rust port of
//! `internal/daemon/uds` (plan §4.2 step 4).
//!
//! [`listen`] ensures the parent dir is `0700`, refuses to take over a socket
//! whose other end accepts connections (a live peer), reaps stale leftovers,
//! binds, and chmods the socket to `0600`. This is what makes the auto-spawn
//! race safe: the loser of a double-spawn gets [`ListenError::AlreadyRunning`]
//! and falls back to connecting to the winner.

use std::io::ErrorKind;
use std::os::unix::fs::PermissionsExt;
use std::os::unix::net::{UnixListener, UnixStream};
use std::path::Path;

/// Why [`listen`] could not bind.
#[derive(Debug)]
pub enum ListenError {
    /// The socket path already has a live peer accepting connections.
    AlreadyRunning,
    /// Any other failure (mkdir / probe / bind / chmod).
    Io(std::io::Error),
}

impl std::fmt::Display for ListenError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ListenError::AlreadyRunning => write!(f, "uds: daemon already running"),
            ListenError::Io(e) => write!(f, "uds: {e}"),
        }
    }
}

impl std::error::Error for ListenError {}

impl From<std::io::Error> for ListenError {
    fn from(e: std::io::Error) -> Self {
        ListenError::Io(e)
    }
}

/// Binds a unix listener at `path` with single-instance semantics:
///
/// 1. create `dir(path)` at mode `0700` (best-effort);
/// 2. if a file exists at `path`, probe it — a successful connect means a live
///    peer ([`ListenError::AlreadyRunning`]); `ECONNREFUSED`/`ENOENT` means a
///    stale leftover, which is unlinked;
/// 3. `bind`, then `chmod 0600`.
pub fn listen(path: &Path) -> Result<UnixListener, ListenError> {
    if let Some(dir) = path.parent() {
        if !dir.as_os_str().is_empty() {
            std::fs::create_dir_all(dir)?;
            let _ = std::fs::set_permissions(dir, std::fs::Permissions::from_mode(0o700));
        }
    }

    if path.symlink_metadata().is_ok() {
        match UnixStream::connect(path) {
            Ok(c) => {
                drop(c);
                return Err(ListenError::AlreadyRunning);
            }
            Err(e) if is_dead_socket(&e) => match std::fs::remove_file(path) {
                Ok(()) => {}
                Err(e) if e.kind() == ErrorKind::NotFound => {}
                Err(e) => return Err(ListenError::Io(e)),
            },
            Err(e) => return Err(ListenError::Io(e)),
        }
    }

    let listener = UnixListener::bind(path)?;
    if let Err(e) = std::fs::set_permissions(path, std::fs::Permissions::from_mode(0o600)) {
        let _ = std::fs::remove_file(path);
        return Err(ListenError::Io(e));
    }
    Ok(listener)
}

/// Removes the socket file (best-effort; `ENOENT` is a no-op).
pub fn cleanup(path: &Path) {
    if let Err(e) = std::fs::remove_file(path) {
        if e.kind() != ErrorKind::NotFound {
            // Nothing actionable on shutdown; the next start reaps it anyway.
        }
    }
}

/// Reports whether a connect error is consistent with a stale leftover (no peer
/// accepting, the file vanished between stat and connect, or the peer closed
/// mid-handshake). Mirrors the Go `isDeadSocketErr`, which reaps
/// `ECONNREFUSED` / `ENOENT` / `ECONNRESET`.
fn is_dead_socket(e: &std::io::Error) -> bool {
    matches!(
        e.kind(),
        ErrorKind::ConnectionRefused | ErrorKind::NotFound | ErrorKind::ConnectionReset
    )
}
